kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -284,6 +372,333 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+  val simpleName = kotlinMetadataName.substringAfterLast(
+      '/', // Drop the package name, e.g. "package/of/class/"
+      '.' // Drop any enclosing classes, e.g. "MyClass."
+  )
+  val packageName = kotlinMetadataName.substringBeforeLast("/", missingDelimiterValue = "")
+  val simpleNames = kotlinMetadataName.removeSuffix(simpleName)
+      .removeSuffix(".") // Trailing "." if any
+      .removePrefix(packageName)
+      .removePrefix("/")
+      .let {
+        if (it.isNotEmpty()) {
+          it.split(".")
+        } else {
+          // Don't split, otherwise we end up with an empty string as the first element!
+          emptyList()
+        }
+      }
+      .plus(simpleName)
+
+  return "${packageName.replace("/", ".")}.${simpleNames.joinToString(".")}"
+}
+
+@Suppress("SameParameterValue")
+private fun String.substringAfterLast(vararg delimiters: Char): String {
+  val index = lastIndexOfAny(delimiters)
+  return if (index == -1) this else substring(index + 1, length)
+}
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+    if (type is Class<*> && type.isPrimitive) {
+      when (type) {
+        Boolean::class.java -> false
+        Char::class.java -> 0.toChar()
+        Byte::class.java -> 0.toByte()
+        Short::class.java -> 0.toShort()
+        Int::class.java -> 0
+        Float::class.java -> 0f
+        Long::class.java -> 0L
+        Double::class.java -> 0.0
+        Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+        else -> throw UnsupportedOperationException("Unknown primitive: $type")
+      }
+    } else null
+
+private fun Class<*>.toKmClass(throwOnNotClass: Boolean): KmClass? {
+  val metadata = getAnnotation(KOTLIN_METADATA) ?: return null
+  val header = with(metadata) {
+    KotlinClassHeader(
+        kind = kind,
+        metadataVersion = metadataVersion,
+        bytecodeVersion = bytecodeVersion,
+        data1 = data1,
+        data2 = data2,
+        extraString = extraString,
+        packageName = packageName,
+        extraInt = extraInt
+    )
+  }
+  val classMetadata = KotlinClassMetadata.read(header)
+  if (classMetadata !is KotlinClassMetadata.Class) {
+    if (throwOnNotClass) {
+      throw IllegalStateException("Cannot serialize class with metadata $classMetadata")
+    } else {
+      return null
+    }
+  }
+
+  return classMetadata.toKmClass()
+}
+
+private class JvmSignatureSearcher(clazz: Class<*>) {
+
+  private val methodSignatureCache = mutableMapOf<String, Method>()
+  private val fieldSignatureCache = mutableMapOf<String, Field>()
+  private val declaredMethodsIterator by lazy(NONE) { clazz.declaredMethods.iterator() }
+  private val declaredFieldsIterator by lazy(NONE) { clazz.declaredFields.iterator() }
+  private val methodIterator by lazy(NONE) { clazz.methods.iterator() }
+  private val fieldIterator by lazy(NONE) { clazz.fields.iterator() }
+
+  fun findMethod(signature: JvmMethodSignature): Method? {
```

Comment:
FWIW, there’s another strategy here . . . decode `signature` into a `Class[]` and then call `getDeclaredMethod`

Reply:
5db8757